proposals/NNNN-remove-sequence-subsequence.md

```diff
@@ -0,0 +1,327 @@
+just adjusting the lower bound upwards), whereas after removing the
+customization points, the `suffix` function that’s called is the one for a
+`Sequence`, which needs to iterate the range and buffer the elements.
+
+This is a nice performance tweak, but it doesn’t justify the significant
+downsides listed above. It is essentially improving generic performance at the
+expense of concrete performance. Normally, these kind of generic improvements
+come at just a tiny cost (e.g. slight increase in compile time or binary size)
+rather than a significant runtime performance penalty.
+
+## Proposed solution
+
+Remove the `SubSequence` associated type from `Sequence`. It should first
+appear from `Collection` onwards.
+
+Remove the methods on `Sequence` that return `SubSequence` from the protocol.
+They should remain as extensions only. Each one should return a specific type
+best suited to the task:
+
+```swift
+extension Sequence {
+  public func dropFirst(_ k: Int = 1) -> DropFirstSequence<Self>
+  public func dropLast(_ k: Int = 1) -> [Element]
+  public func suffix(_ maxLength: Int) -> [Element]
+  public func prefix(_ maxLength: Int) -> PrefixSequence<Self>
+  public func drop(while predicate: (Element) throws -> Bool) rethrows -> DropWhileSequence<Self>
+  public func prefix(while predicate: (Element) throws -> Bool) rethrows -> [Element]
+  public func split(
+    maxSplits: Int = Int.max,
+    omittingEmptySubsequences: Bool = true,
+    whereSeparator isSeparator: (Element) throws -> Bool
+  ) rethrows -> [ArraySlice<Element>]
+}
+```
+
+`DropFirstSequence`, `DropWhileSequence` and `PrefixSequence` already exist in
+the standard library in underscored form.
+
+This will also have the useful side-effect that these methods can also be
+removed as customization points from `Collection` as well, similar to removing
+`prefix(upTo:)` in
+[SE-0232](https://github.com/apple/swift-evolution/blob/master/proposals/0232-re
+move-customization-points.md), because there’s no longer any reasonable
+customization to be done on a per-collection basis.
+
+Doing this will have considerable benefits to code size as well. For example,
+the CoreAudio overlay that declares a handful of collections reduces in size by
+25% after applying these changes.
+
+Once done, this change will allow the following simplifying changes and
+enhancements to standard library types:
+
+- `LazySequenceProtocol` and `LazyCollectionProtocol` can be collapsed into a
+  single protocol.
+
+- The following type can be collapsed, dropping the collection variant (with a
+  typealias provided for source compatability):
+
+    - `FlattenSequence` and `Collection`
+    - `LazySequence` and `LazyCollection`
+    - `LazyMapSequence` and `Collection`
+    - `LazyFilterSequence` and `Collection`
+    - `LazyDropWhileSequence` and `Collection`
+    - `LazyPrefixWhileSequence` and `Collection`
+    - `LazyMapSequence` and `Collection`
+    - `FlattenSequence` and `Collection`
```

Comment:
The last two lines are duplicates.

It might be clearer to write the collection variants in full, or at least prefix them with a hyphen.

- `FlattenSequence` and `FlattenCollection`

*or*

- `FlattenSequence` and `-Collection`